lib/farscape/plugins.rb

```diff
@@ -2,59 +2,91 @@
 
 module Farscape
 
-  @plugins = []
-  @disabled_plugins = []
-  @middleware_stack = PartiallyOrderedList.new { |m,n| order_middleware(m,n) }
+  @plugins = {}
+  @disabling_rules = []
 
   class <<self
 
     attr_reader :plugins
-    attr_reader :disabled_plugins
-    attr_reader :middleware_stack
+    attr_reader :disabling_rules
 
     def register_plugin(options)
-      return false if disabled?(options)
-      @plugins << options
-      add_middleware(options)
-      true
+      options[:enabled] = enabled?(options)
+      @plugins[options[:name]] = options
+    end
+
+    def register_plugins(a_list)
+      a_list.each { |options| register_plugin(options) }
+    end
+
+    def enabled_plugins
+      @plugins.select { |plugin| @plugins[plugin][:enabled] }
+    end
+
+    def disabled_plugins
+      @plugins.reject { |plugin| @plugins[plugin][:enabled] }
     end
 
     # If the middleware has been disabled by name, return the name
     # Else if by type, return the type.
+    # Else if :default_state was passed in return :default_state
+    def why_disabled(options)
+      maybe = @disabling_rules.map { |hash| hash.select { |k,v| k if v == options[k] } }
+      maybe |= disabled_plugins.map { |k,v| v[:name] }
+      maybe |= [:default_state] if options[:default_state] == :disabled
+      maybe || []
+    end
+    
     def disabled?(options)
-      (disabled_plugins & [options[:name],options[:type]]).first
+      options = normalize_selector(options)
+      return @plugins[options[:name]][:enabled] if options.include?([:name])
+      why_disabled(options).any?
+    end
+    
+    def enabled?(options)
+      !disabled?(options)
     end
 
-    # Prevents a plugin [type] from being registered, and unregisters it if it's already there
-    def disable(name_or_type)
-      disabled_plugins << name_or_type
-      plugins_to_disable, @plugins = @plugins.partition { |plugin| [plugin[:name], plugin[:type]].include?(name_or_type) }
-      plugins_to_disable.each { |plugin| disable_plugin(plugin) }
+    # Prevents a plugin from being registered, and disables it if it's already there
+    def disable!(name_or_type)
+      name_or_type = normalize_selector(name_or_type)
+      @disabling_rules << name_or_type
+      selected_plugins = find_attr_intersect(@plugins, name_or_type)
+      selected_plugins.each { |plugin| @plugins[plugin][:enabled] = false }
     end
 
+    # Allows a plugin to be registered, and enables it if it's already there
+    def enable!(name_or_type)
+      name_or_type = normalize_selector(name_or_type)
+      @disabling_rules.delete(name_or_type)
+      selected_plugins = find_attr_intersect(@plugins, name_or_type)
+      selected_plugins.each { |plugin| @plugins[plugin][:enabled] = true }
+    end
+
+    # Returns the Poset representing middleware dependency
+    def middleware_stack
```

Comment:
Could we cache the result of computing the `middleware list` and reference that here and then update the list as when new middlewares are added and/or removed. This way we keep making http calls as streamlined as possible.